client/beefy/src/worker.rs

```diff
@@ -216,77 +262,211 @@ where
+
+				info!(target: "beefy", "🥩 Round #{} concluded, committed: {:?}.", round.1, signed_commitment);
+
+				if let Err(e) = self.backend.append_justification(
+					BlockId::Number(block_num),
+					(
+						BEEFY_ENGINE_ID,
+						VersionedFinalityProof::V1(signed_commitment.clone()).encode(),
+					),
+				) {
+					trace!(target: "beefy", "🥩 Error {:?} on appending justification: {:?}", e, signed_commitment);
+				}
+				self.signed_commitment_sender
+					.notify(|| Ok::<_, ()>(signed_commitment))
+					.expect("forwards closure result; the closure always returns Ok; qed.");
+
+				self.set_best_beefy_block(block_num);
+
+				trace!(target: "beefy", "🥩 round concluded, waking up ticker");
+				self.ticker.wake();
+			}
+		}
+	}
+
+	/// Set best BEEFY block to `block_num`.
+	///
+	/// Also sends/updates the best BEEFY block hash to the RPC worker.
+	fn set_best_beefy_block(&mut self, block_num: NumberFor<B>) {
+		if Some(block_num) > self.best_beefy_block {
+			// Try to get block hash ourselves.
+			let block_hash = match self.client.hash(block_num) {
+				Ok(h) => h,
+				Err(e) => {
+					error!(target: "beefy", "🥩 Failed to get hash for block number {}: {}",
+						block_num, e);
+					None
+				},
+			};
+			// Update RPC worker with new best BEEFY block hash.
+			block_hash.map(|hash| {
+				self.beefy_best_block_sender
+					.notify(|| Ok::<_, ()>(hash))
+					.expect("forwards closure result; the closure always returns Ok; qed.")
+			});
+			// Set new best BEEFY block number.
+			self.best_beefy_block = Some(block_num);
+			metric_set!(self, beefy_best_block, block_num);
+		} else {
+			debug!(target: "beefy", "🥩 Can't set best beefy to older: {}", block_num);
+		}
+	}
+
+	/// Return first block number pertaining to same session as `header`.
+	fn session_start_for_header(&self, header: &B::Header) -> NumberFor<B> {
+		// Current header is also the session start.
+		if let Some(_) = find_authorities_change::<B>(header) {
+			return *header.number()
+		}
+
+		// Walk up the chain looking for the session change digest.
+		let mut header = header.clone();
+		let mut parent_hash = *header.parent_hash();
+		while parent_hash != Default::default() {
+			header = self
+				.client
+				.expect_header(BlockId::Hash(parent_hash))
+				// TODO: is this proof correct?
+				.expect("header always available when following parent hash; qed.");
+			if let Some(_) = find_authorities_change::<B>(&header) {
+				// Found the first header of the session.
+				info!(target: "beefy", "🥩 session boundary is: {:?}.", *header.number());
+				return *header.number()
+			}
+			parent_hash = *header.parent_hash();
+		}
+
+		info!(target: "beefy", "🥩 no session boundary found, defaulting to block number 1.");
+		1u32.into()
+	}
+
+	/// Handle potential session changes by starting new voting round for mandatory blocks.
+	fn handle_session_change(&mut self) {
+		let header = &self.best_grandpa_block_header;
+		if let Some(active) = self.validator_set(header) {
```

Comment:
Continuing on the comment above `validator_set` will call into `find_authorities_change` and then fallback to calling into the runtime if no authority change digest exists on the block. IMO we should just rely on the header digest since we already assume an invariant that GRANDPA will always finalize blocks that change the validator set.

Reply:
Sorry I forgot to mention, this was recently changed with https://github.com/paritytech/substrate/pull/10639! :) So now only one finality notification is sent for all the blocks we finalize and the notification itself includes all the intermediary blocks that were finalized, i.e. if current finalized block is #10 and there is a call `client.finalize_block(#15)` you will receive a single finality notification which itself says that block #15 was finalized and blocks #11, #12, #13, #14 were also finalized. The only scenario where we don't send any notifications is when we are doing the initial major sync.

Reply:
So for example is the following scenario possible (looking at that PR, it seems yes)?

- session-length is 10 blocks,
- beefy worker gets finality notification for `#7` (implying all other <7 also finalized),
- validators + session change at block `#11`
- beefy worker gets finality notification for `#12` (includes list for `#8 #9 #10 #11` as finalized blocks)

In this case if we only process `best_grandpa` (`#12`) we miss the _new validator set_ digest in header `#11`.

In such a case, the current PR finds out about the new session/validator-set through runtime `BeefyApi` call. If it detects new session it walks back the header chain looking for the digest.

Reply:
So first thing I think we would always get a finality notification for `#11` since GRANDPA is guaranteed to finalize that block and thus it will result in a call to `client.finalize_block(#11)`. But regardless of that, we can relax this invariant and instead always check the digests of intermediary finalized blocks as you suggested (maybe we can have some edge condition where if the number of intermediaries is > 1024 e.g. we resort to calling into the runtime as we do on the startup case, but I think we can worry about that later).